[Explicit Module Builds] Switch versioned `canImport` to return `true` when encountering unversioned candidate

artemcm opened the pull request:
Specifically, when the scanner found a candidate which does not carry a user-specified version, it will pass `-module-can-import Foo` to compilation. During compilation, if the check is versioned but the candidate is unversioned, evaluate the check to 'true' to restore the behavior we had with implicitly-built modules.

Resolves rdar://148134993

artemcm committed:
[Explicit Module Builds] Switch versioned 'canImport' to return 'true' when encountering unversioned candidate

Specifically, when the scanner found a candidate which does not carry a user-specified version, it will pass '-module-can-import Foo' to compilation. During compilation, if the check is versioned but the candidate is unversioned, evaluate the check to 'true' to restore the behavior we had with implicitly-built modules.

Resolves rdar://148134993

cachemeifyoucan commented:
Can you extend tests in `ScanDependencies/module_deps_can_import.swift` to make sure the scanner returns `-module-can-import` in this case?

artemcm committed:
[Dependency Scanning] Always record best version of discovered 'canImport'-ed modules

Suppose module 'Foo' exists in the search paths and specifies user module version '1.0'.

If the first encountered 'canImport' query is unversioned:
...

Followed by a versioned one:
...

The success of the first check will record an unversioned successful canImport, which will cause the second check to evaluate to 'true', which is incorrect.

This change causes even unversioned 'canImport' checks to track and record the discovered user module version.

artemcm commented:
@swift-ci smoke test

artemcm commented on lib/AST/ASTContext.cpp:
```diff
@@ -2709,15 +2715,43 @@ bool ASTContext::canImportModuleImpl(ImportPath::Module ModuleName,
+    // and record the module version that the succeeding loader
+    // observed.
+    ModuleLoader::ModuleVersionInfo bestVersionInfo;
+    bool moduleFound = false;
```
This was a second, unrelated by that I found when implementing the titular change, thanks to our test suite.

Suppose module `Foo` exists in the search paths and specifies user module version `1.0`.
If the first encountered `canImport` query is unversioned:
```
#if canImport(Foo)
...
#endif
```

Followed by a versioned one:
```
#if canImport(Foo, _version: 2.0)
...
#endif
```

The success of the first check will record an unversioned successful `canImport`, which will cause the second check to evaluate to `true`, which is incorrect.

This change causes even unversioned `canImport` checks to track and record the discovered user module version.

cachemeifyoucan reviewed (commented):
LGTM in general.

Just to make sure I followed correctly. Now all `canImport` check on versioned module will result in `-module-can-import-version` and only un-versioned module will get `-module-can-import`, right?

My only concern is we should avoid parsing TBD file if it is not underlying version.

artemcm commented:
> Just to make sure I followed correctly. Now all `canImport` check on versioned module will result in `-module-can-import-version` and only un-versioned module will get `-module-can-import`, right?

That's right, yeah.

artemcm commented:
> My only concern is we should avoid parsing TBD file if it is not underlying version.

Is your concern about performance? We do not avoid parsing TBDs in the plain versioned lookup either today. Though in theory we could guard the clang-specific `::canImportModule` call to not record/return a version if it is not an underlying version lookup?

cachemeifyoucan commented:
> Is your concern about performance?

Yeah, as we don't parse TBD file when the check is not versioned. It might be fine as the cost only incurs with canImport and is proportional to the number of canImport statements. I will let you decide if the trouble is worth it.

artemcm commented:
@swift-ci smoke test

cachemeifyoucan commented on lib/AST/ASTContext.cpp:
```diff
@@ -2692,6 +2692,12 @@ bool ASTContext::canImportModuleImpl(ImportPath::Module ModuleName,
       return true;
 
+    // A versioned canImport, upon encountering an unversioned module
+    // for the queried identifier, evaluates to `true`.
```
Should we only check one version based on `underlyingVersion`?

For example, an swift versioned canImport check should still succeed if there is a clang module overlay for the module which has underlying version?

artemcm replied:
Good catch, thank you. Fixed.

artemcm commented:
@swift-ci smoke test

artemcm commented:
@swift-ci smoke test

artemcm commented:
@cachemeifyoucan I've updated the PR with a more-thorough refactor of `canImportModuleImpl` which now always records both the primary and underlying versions of all modules it discovers. My main goal here was to ensure:
1. As per the title, have versioned `canImport` return `true` when all we find is an un-versioned module candidate.
2. When compilation has `-module-can-import` command-line flags in EBM mode, it never goes on to try and look up modules.
3. Outside scanning or EBM compilation, during an implicit build, encountering a `canImport` of any module will result in populating an entry for it in `CanImportModuleVersions` which will contain both the main module version and underlying, if one exists, and will be re-used for all future `canImport` queries of this module, versioned and un-versioned.  

 I'd appreciate another look at these changes.

cachemeifyoucan commented on lib/AST/ASTContext.cpp:
```diff
@@ -2692,35 +2692,99 @@ bool ASTContext::canImportModuleImpl(ImportPath::Module ModuleName,
-    } else {
-      if (!Found->second.Version.empty())
-        return version <= Found->second.Version;
+    auto foundComparisonVersion = underlyingVersion
```
`auto &` to avoid copy?

cachemeifyoucan commented on lib/AST/ASTContext.cpp:
```diff
@@ -2648,34 +2646,22 @@ static bool isClangModuleVersion(const ModuleLoader::ModuleVersionInfo &info) {
-    else
-      entry.Version = versionInfo;
-  }
+  if (!versionInfo.empty())
```
Maybe the check is unnecessary now?

cachemeifyoucan commented on include/swift/AST/ASTContext.h:
```diff
@@ -1184,12 +1184,14 @@ class ASTContext final {
-                           llvm::VersionTuple &foundVersion) const;
+                           bool isSourceCanImport,
+                           llvm::VersionTuple &foundVersion,
+                           llvm::VersionTuple &foundUnderlyingClangVersion) const;
```
I feel like if we just sink `addSucceededCanImportModule` into `canImportModuleImpl`, there is no need to pass the parameter here. It made the code a bit cleaner that way.

cachemeifyoucan replied:
I guess I did found we need two versions of the can import check with only one of them updating the mapping so the alternative is passing a flag. Not that much difference so it might be fine.

artemcm replied:
Yeah I share the sentiment, but it's a bit more trouble than it is worth, I think, since there's some value in having this function be `const`. And on top of that, there are multiple `return true` code-paths here and if we ever added new ones, we'd need to ensure that we call `addSucceededCanImportModule` on it before existing, which is trickier to enforce.

cachemeifyoucan commented on lib/AST/ASTContext.cpp:
```diff
@@ -2692,35 +2692,99 @@ bool ASTContext::canImportModuleImpl(ImportPath::Module ModuleName,
+      if (!importer->canImportModule(ModuleName, loc, &versionInfo))
+        continue; // The loader can't find the module.
+
+      if (validateVersion(bestVersionInfo, versionInfo,
```
Doesn't `canImportModule` only return one version, it is either clang version (underlying) or swift version?

So you only need to validateVersion once depending on the kind of version returned from importer.

artemcm replied:
I think the subtlety here is that a Clang dependency module version may be not just the underlying version but also the main lookup version. If all that exists for the queried module is a Clang module with that name, and no Swift module at all, then the non-underlying version will be set to the same value as the underlying version. So we need to validate returned version from `canImportModule` against both the currently-known best version and currently-known best underlying version.

cachemeifyoucan reviewed (changes requested):
Some comments about coding. Function-wise, it is good.

artemcm commented:
@swift-ci test

cachemeifyoucan commented:
LGTM. Thanks!
